diem-move/diem-framework/experimental/sources/NFT.move

```diff
@@ -84,62 +79,83 @@ address 0x1 {
-            &wrapper.metadata
+        /// Returns the overall supply for the given token
+        public fun get_supply<TokenType: copy + store + drop>(token: &Token<TokenType>): u64 acquires TokenDataCollection{
+            if (Option::is_some(&token.token_data)) {
+                Option::borrow(&token.token_data).supply
+            } else {
+                let creator_addr = GUID::id_creator_address(&token.id);
+                let creator_tokens_data = &borrow_global<TokenDataCollection<TokenType>>(creator_addr).tokens;
+                let token_data_idx = *Option::borrow(&index_of_token<TokenType>(creator_tokens_data, &token.id));
+                Vector::borrow(creator_tokens_data, token_data_idx).supply
+            }
         }
 
-        /// Returns ID of collection associated with token
-        public fun parent<TokenType: store>(wrapper: &TokenDataWrapper<TokenType>): &Option<GUID::ID> {
-            &wrapper.parent_id
+        /// Returns a copy of the token content uri
+        public fun get_content_uri<TokenType: copy + store + drop>(token: &Token<TokenType>): vector<u8> acquires TokenDataCollection {
+            if (Option::is_some(&token.token_data)) {
+                *&Option::borrow(&token.token_data).content_uri
+            } else {
+                let creator_addr = GUID::id_creator_address(&token.id);
+                let creator_tokens_data = &borrow_global<TokenDataCollection<TokenType>>(creator_addr).tokens;
+                let token_data_idx = *Option::borrow(&index_of_token<TokenType>(creator_tokens_data, &token.id));
+                *&Vector::borrow(creator_tokens_data, token_data_idx).content_uri
+            }
         }
 
-        /// Returns the supply of tokens with `id` on the chain.
-        public fun supply<TokenType: store>(id: &GUID::ID): u64 acquires TokenDataCollection {
-            let owner_addr = GUID::id_creator_address(id);
-            let tokens = &mut borrow_global_mut<TokenDataCollection<TokenType>>(owner_addr).tokens;
-            let index_opt = index_of_token<TokenType>(tokens, id);
-            assert!(Option::is_some(&index_opt), Errors::invalid_argument(EWRONG_TOKEN_ID));
-            let index = Option::extract(&mut index_opt);
-            Vector::borrow(tokens, index).supply
+        /// Returns a copy of the token metadata
+        public fun get_metadata<TokenType: copy + store + drop>(token: &Token<TokenType>): TokenType acquires TokenDataCollection {
+            if (Option::is_some(&token.token_data)) {
+                *&Option::borrow(&token.token_data).metadata
+            } else {
+                let creator_addr = GUID::id_creator_address(&token.id);
+                let creator_tokens_data = &borrow_global<TokenDataCollection<TokenType>>(creator_addr).tokens;
+                let token_data_idx = *Option::borrow(&index_of_token<TokenType>(creator_tokens_data, &token.id));
+                *&Vector::borrow(creator_tokens_data, token_data_idx).metadata
```

Comment:
Probably could have put this whole blob in a separate function with `.variable`. Though move might make it hard

Reply:
Yea, I wish :). We can't put it in a separate function, because functions are not allowed to return references to global storage.

Reply:
Looks like we'll need macros sometime